### DISCUSSION TITLE
Secrets in ConfigurableResources

### DISCUSSION THREAD

**Discussion by matthias-Q** [dagster-io/dagster#18829] [Q&A]
Hi,
I am trying to understand how ConfigurableResources should be handled with secrets.

I want to use the `SSHResource` and I want to provide the ssh key as a string. This is fetched from the AWS secrets store.
Since ConfigurableResources are now immutable (derived from pydantic.BaseModel with froze=True) I have to provide this secret to the `SSHResource` in the `Definition`.

This means it is in the `__init__.py` in my user_code and the secret is loaded when the user_code is loaded and not at the beginning of the execution. Is this the desired way? I have issues with mocking things for testing in CI when the code is already in the __init__.py.

What is the correct way here?

**Comment by sventech** [comment]
I'm also facing this issue and am not testing some of the code I'd like to.

**Comment by ion-elgreco** [comment]
Generally fetching them at runtime and then passing it into a configurable resource is fine. But if they are short lived and you want to lazy call/set this. You can set a method `setup_for_execution`https://docs.dagster.io/guides/build/external-resources/managing-resource-state to set the tokens or secrets when the resource is constructed
